pkg/pool-utils/contracts/RestrictedBalancerPoolToken.sol

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+
+import "./BalancerPoolToken.sol";
+import "./BasePoolAuthorization.sol";
+
+/**
+ * @title Extension of BalancerPoolToken with holder allowlist
+ * @author Balancer Labs
+ * @dev
+ * - Restricts both senders and recipients of all transfers (including mints
+ *   and burns) to addresses on an allowlist
+ * - Includes permissioned functions for adding/removing members to/from the
+ *   allowlist
+ */
+abstract contract RestrictedBalancerPoolToken is BalancerPoolToken, BasePoolAuthorization {
+    mapping(address => bool) private _allowedAddresses;
+
+    event AddressAdded(address member);
+    event AddressRemoved(address member);
+
+    constructor(string memory tokenName, string memory tokenSymbol, address owner)
+        BalancerPoolToken(tokenName, tokenSymbol)
+        BasePoolAuthorization(owner)
+    {
+        // Add zero address to allowlist to avoid blocking mints and burns
+        _allowedAddresses[address(0)] = true;
```

Comment:
With the other changes this would be unnecessary, resulting in overall less gas usage (and I'd argue more robust code, as e.g. the zero address could never be removed from the list of allowed addresses).

Reply:
Fair point about removing the zero address!